test/logs_test.go

```diff
@@ -0,0 +1,145 @@
+package telepilot_test
+
+import (
+	"bytes"
+	"context"
+	"os"
+	"sync"
+	"testing"
+
+	pb "go.creack.net/telepilot/api/v1"
+)
+
+func TestStreamLogsSimple(t *testing.T) {
+	t.Parallel()
+
+	ts, ctx := newTestServer(t)
+
+	t.Run("no output", func(t *testing.T) {
+		t.Parallel()
+
+		// Create a Job without output.
+		jobID, err := ts.bob.StartJob(ctx, "true", nil)
+		noError(t, err, "Bob start job.")
+
+		// Stream logs and assert.
+		w := bytes.NewBuffer(nil)
+		noError(t, ts.bob.StreamLogs(ctx, jobID, w), "Stream Logs.")
+		assert(t, "", w.String(), "invalid output")
+
+		// One more time should yield same result.
+		w.Reset()
+		noError(t, ts.bob.StreamLogs(ctx, jobID, w), "Stream Logs.")
+		assert(t, "", w.String(), "invalid output")
+	})
+
+	t.Run("finite output", func(t *testing.T) {
+		t.Parallel()
+
+		// Create a Job with a few lines.
+		jobID, err := ts.bob.StartJob(ctx, "sh", []string{"-c", "echo hello; echo world"})
+		noError(t, err, "Bob start job.")
+
+		// Stream logs and assert.
+		w := bytes.NewBuffer(nil)
+		noError(t, ts.bob.StreamLogs(ctx, jobID, w), "Stream Logs.")
+		assert(t, "hello\nworld\n", w.String(), "invalid output")
+
+		// Assert that the process is indeed exited.
+		status, err := ts.bob.GetJobStatus(ctx, jobID)
+		noError(t, err, "Get Job Status.")
+		assert(t, pb.JobStatus_JOB_STATUS_EXITED.String()+" (0)", status, "unexpected status after stream logs")
+
+		// One more time should yield same result.
+		// The previous time the output may be from live feed, this time, we know it is historical data.
+		w.Reset()
+		noError(t, ts.bob.StreamLogs(ctx, jobID, w), "Stream Logs.")
+		assert(t, "hello\nworld\n", w.String(), "invalid output")
+	})
+}
+
+// Test streaming logs of a long-running process.
+func TestStreamLogsOngoingOutput(t *testing.T) {
+	t.Parallel()
+
+	ts, ctx := newTestServer(t)
+
+	// Waitgroup to make sure all routines are done at the end.
+	var wg sync.WaitGroup
+	t.Cleanup(wg.Wait)
+
+	// Pipe to sync the job with the test.
+	pipeName := mkTempPipe(t, "pipe")
+
+	// Create a Job with a few lines and keep the job alive.
+	jobID, err := ts.bob.StartJob(
+		ctx,
+		"sh", []string{"-c", "cat " + pipeName + "; echo hello; echo world; cat " + pipeName},
```

Comment:
The is fine, but I would be tempted to just run a shell loop with a short sleep (or some other similar-behaving program). You get to do the same assertions but cut all the pipe logic.

(No need to change the tests.)

Reply:
I thought about this, but simply sleeping would either slow down the tests a lot or cause instability.

Once (if) we implement support for input, we can use stdin instead of a pipe.

I'll add a TODO.

Reply:
TODO added on the mkTempPipe helper.